### ISSUE TITLE
dynamic cache/memtable reservations

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#3522]
Currently, we reserve 50% of memory to memtables and 50% to cache. This provides a good balance between read workloads that want a large cache and write workloads that want large memtables to increase write coalescing and reduce write amplification.

We can improve this by dynamically adjusting the ratio according to the workload. A read-intensive workload can use a 80%/20% ratio to increase cache hit rates. Increased write amplification will be tolerable, since the overall write amplification is low. Similarly, write-intensive workloads can benefit from larger memtables, so they'll use a 20%/80% reservation; while reads will become more expensive the low read rate means this is offset by the improvement in writes.

Switching between reservations should be done very gently to avoid hitting the wall (when decreasing memtable reservations) or causing an eviction spree (when decreasing cache reservations). 80%/20% should be reached when the read/write ratio is at least 10:1.

**Comment by avikivity** [comment]
We can prototype this by changing the reservations statically and testing pure read and pure write workloads.

**Comment by dorlaor** [comment]
Good idea. Reads portion should also depends on the relative cache-ability
of it. If you read a lot but within the working set,
there is no need to increase it. If you read with random access pattern,
additional memory for cache won't help.

But adjusting the memory dynamically will be good before we go through the
working set size estimation and eviction.
